Objects/regions.c

```diff
@@ -92,9 +116,324 @@ struct regionmetadata {
     regionmetadata* next;
 };
 
-static bool is_bridge_object(PyObject *op);
-static int regionmetadata_has_ancestor(regionmetadata* data, regionmetadata* other);
-static regionmetadata* PyRegion_get_metadata(PyRegionObject* obj);
+static Py_region_ptr_t regionmetadata_get_root(Py_region_ptr_t self)
+{
+    // Test for constants
+    if (IS_REGION_NULL(self) || IS_CONST_REGION(self)) {
+        return _Py_CAST(Py_region_ptr_t, self);
+    }
+
+    // Return self if it wasn't merged with another region
+    regionmetadata* self_data = _Py_CAST(regionmetadata*, self);
+    if (!REGION_PTR_GET_TAG(self_data->parent, Py_METADATA_MERGE_TAG)) {
+        return self;
+    }
+
+    // FIXME: It can happen that there are several layers in this union-find
+    // structure. It would be efficient to directly update the parent pointers
+    // for deeper nodes.
+    return regionmetadata_get_root(Py_region_ptr(self_data->parent));
```

Comment:
I don't think we want a tail recursive version (because C and optimisations…). Can this be rewritten with a loop instead?

Reply:
Probably. But I expect to rewrite this soon anyways for `O(1)` merging. The `try_close` PR will only merge stuff into the local region, thus creating a maximum depth of 1